### PR TITLE
Check for correct configuration of Envoy listener for a given Pod

### PR DESCRIPTION
**Pull request by draychev**
This PR adds a very basic check for Envoy listener -- answers the question - does "outbound-listener" exist.

### CODE REVIEW

File: pkg/envoy/config.go [hunk 1]
```diff
@@ -1,9 +1,11 @@
...
+	v1 "k8s.io/api/core/v1"
+)
 
-// GetEnvoyConfig returns the Envoy config for the given pod.
-func GetEnvoyConfig(pod k8s.Pod) (*Config, error) {
+// GetEnvoyConfigGetterForPod returns a ConfigGetter struct, which can fetch the Envoy config for the given pod.
+func GetEnvoyConfigGetterForPod(pod *v1.Pod) (ConfigGetter, error) {
 	// TODO(draychev): Get the config from the Envoy sidecar of the given pod.
 	return nil, nil
 }
```
Comment (draychev) [inline pkg/envoy/config.go:11]: @johnsonshi Once you have the code to fetch Envoy config -- we'll reuse that function here to fetch config for a given pod.

File: pkg/envoy/config.go [hunk 2]
```diff
@@ -1,9 +1,29 @@
...
+func (mcg tempStruct) GetObjectName() string {
+	return fmt.Sprintf("%s/%s", mcg.Pod.Namespace, mcg.Pod.Name)
+}
+
+// GetEnvoyConfigGetterForPod returns a ConfigGetter struct, which can fetch the Envoy config for the given pod.
+func GetEnvoyConfigGetterForPod(pod *v1.Pod) (ConfigGetter, error) {
 	// TODO(draychev): Get the config from the Envoy sidecar of the given pod.
-	return nil, nil
+	return tempStruct{Pod: pod}, nil
 }
```
Comment (draychev) [inline pkg/envoy/config.go:29]: @johnsonshi Once you have the code to fetch Envoy config -- we'll reuse that function here to fetch config for a given pod.

Comment (johnsonshi) [reply pkg/envoy/config.go:29]: We'll be using `openservicemesh/osm/pkg/cli`'s `GetEnvoyProxyConfig` for this.

Comment (draychev) [reply pkg/envoy/config.go:29]: @johnsonshi Exactly right - here is where #19 is augmented w/ the new function: https://github.com/openservicemesh/osm-health/pull/25 (WIP)

**Review by johnsonshi — APPROVED** [review]

**Review by SanyaKochhar — APPROVED** [review]

File: pkg/envoy/lds.go [hunk 3]
```diff
@@ -1,41 +1,64 @@
...
+	if envoyConfig == nil || envoyConfig.Listeners.DynamicListeners == nil {
+		return ErrEnvoyConfigEmpty
+	}
+
+	actualListener := envoyConfig.Listeners.DynamicListeners[0]
+
+	if expectedListenerName != actualListener.Name {
+		log.Error().Msgf("must have listener with name %s but it is instead %s", expectedListenerName, actualListener.Name)
+		return ErrEnvoyListenerMissing
+	}
```
Comment (johnsonshi) [inline pkg/envoy/lds.go:48]: Hey @draychev just for my understanding, in what situations does this case arise? If the pod has an envoy sidecar created through some other means (not by our mesh), but our mesh's own envoy sidecar in this pod failed to set up properly, then we have a case of mismatched envoy listener name?

Comment (draychev) [reply pkg/envoy/lds.go:48]: Good question -- Each source Envoy should have an outbound listener and each destination envoy should have an inbound listener.  take a look at this:  https://github.com/openservicemesh/osm-health/pull/25/files#diff-ec2be6376364a5c456575f111cccd27aea580f1b42d26bd4227a484ca1bb9c86R50-R51

**Pull request merged**